refactor(runtime): convert dist#ft functions to lua

smjonas opened the pull request:
Working on #18241, supersedes #18244

gpanders commented on runtime/lua/vim/filetype.lua:
```diff
@@ -929,10 +929,10 @@ local filename = {
-  fvSolution = function() vim.fn["dist#ft#FTfoam"]() end,
-  fvConstraints = function() vim.fn["dist#ft#FTfoam"]() end,
-  fvModels = function() vim.fn["dist#ft#FTfoam"]() end,
+  fvSchemes = function() require("vim.filetype.detect").foam() end,
+  fvSolution = function() require("vim.filetype.detect").foam() end,
+  fvConstraints = function() require("vim.filetype.detect").foam() end,
+  fvModels = function() require("vim.filetype.detect").foam() end,
```
These should all be `return require("vim.filetype.detect").foam()`

gpanders commented on runtime/lua/vim/filetype.lua:
```diff
@@ -1428,14 +1428,14 @@ local pattern = {
-  [".*/constant/g"] = function() vim.fn["dist#ft#FTfoam"]() end,
-  [".*/0/.*"] = function() vim.fn["dist#ft#FTfoam"]() end,
-  [".*/0%.orig/.*"] = function() vim.fn["dist#ft#FTfoam"]() end,
+  ["[a-zA-Z0-9].*Dict"] = function() require("vim.filetype.detect").foam() end,
+  ["[a-zA-Z0-9].*Dict%..*"] = function() require("vim.filetype.detect").foam() end,
+  ["[a-zA-Z].*Properties"] = function() require("vim.filetype.detect").foam() end,
+  ["[a-zA-Z].*Properties%..*"] = function() require("vim.filetype.detect").foam() end,
+  [".*Transport%..*"] = function() require("vim.filetype.detect").foam() end,
+  [".*/constant/g"] = function() require("vim.filetype.detect").foam() end,
+  [".*/0/.*"] = function() require("vim.filetype.detect").foam() end,
+  [".*/0%.orig/.*"] = function() require("vim.filetype.detect").foam() end,
```
Same here

gpanders commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,31 @@
+-- In that case, it's probably an OpenFOAM file
+function M.foam()
+  local foam_file = false
+  for line = 1, 15 do
+    local current_line = vim.fn.getline(line)
```
Use `vim.api.nvim_buf_get_lines(bufnr, 0, 15, true)` and then iterate over that with `ipairs()` instead of calling `vim.fn.getline` in each iteration. Also modify this function to accept a `bufnr` parameter and pass it in `filetype.lua`.

clason replied:
probably best to copy the `getline` function from `filetype.lua`.

gpanders commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,31 @@
+    if current_line:find("^FoamFile") then
+      foam_file = true
+    elseif foam_file and current_line:find("^%s*object") then
+      vim.cmd("setf foam")
```
Should `return "foam"` instead of setting the filetype directly.

gpanders commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,31 @@
+end
+
+-- Determine if a *.tf file is TF mud client or terraform
+function M.tf()
+  local number_of_lines = vim.fn.line("$")
+  for line = 1, number_of_lines do
+    local current_line = vim.fn.trim(vim.fn.getline(line))
+    if not current_line:find("^[;/ ]") then
+      vim.cmd([[setf terraform]])
+      return
+    end
+    vim.cmd([[setf tf]])
+  end
+end
```
All the comments I made above apply here too, but I know this is not your code :) Ping @dundargoc.

clason commented:
Also, link to #18241 to avoid duplication of effort.

smjonas commented on runtime/lua/vim/filetype.lua:
```diff
@@ -929,10 +929,10 @@ local filename = {
-  fvSolution = function() vim.fn["dist#ft#FTfoam"]() end,
-  fvConstraints = function() vim.fn["dist#ft#FTfoam"]() end,
-  fvModels = function() vim.fn["dist#ft#FTfoam"]() end,
+  fvSchemes = function(path, bufnr) return require("vim.filetype.detect").foam(bufnr) end,
```
Should I rename the unused `path` argument to `_`?

clason replied:
No; please keep to the patterns in `filetype.lua` -- things should be consistent; they can still be improved (en masse) afterwards.

clason commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,29 @@
+function M.tf(path, bufnr)
+  local number_of_lines = vim.fn.line("$")
+  for line = 1, number_of_lines do
+    local current_line = vim.fn.trim(vim.fn.getline(line))
```
Would be good to replace this with a local convenience function that takes a `bufnr`(!) and uses `vim.api.nvim_buf_line_count(bufnr)` instead of `getline`. (`trim` is probably also replaceable by a pure Lua function; maybe something like

```lua
local function trim(s)
   local from = s:match"^%s*()"
   return from > #s and "" or s:match(".*%S", from)
end
```

gpanders replied:
Note that we have `vim.trim` as a Lua function already.

clason replied:
Oh, nice, I did think that was too useful not to have as a standard library function ;)

smjonas committed:
refactor(runtime): convert dist#ft functions to lua

smjonas committed:
Rename getline to get_lines and use it instead of nvim API functions

smjonas committed:
Add functions for nroff, redif; use get_lines correctly

smjonas committed:
Use iter_lines helper function, add 4 more checks

smjonas committed:
Add xml function

smjonas committed:
Add functions for btm, ent, progress_cweb, sc, scd and y

smjonas committed:
Sort functions alphabetically, add skeleton functions, remove _check suffix

clason commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,362 @@
+end
+
+function M.btm(_, bufnr)
+  if vim.g.dosbatch_syntax_for_btm ~= nil and vim.g.dosbatch_syntax_for_btm ~= 0 then
```
```suggestion
  if vim.g.dosbatch_syntax_for_btm ~= 0 then
```
(`0` is truthy in Lua, and `nil ~=0`)

smjonas replied:
But if it is `nil`, shouldn't that set the filetype to `btm`? That won't happen if we omit the first check here.

clason replied:
Ah, you're right, I missed that (sorry!)

You can still drop the `~=nil`, though :)

clason commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,362 @@
+end
+
+function M.euphoria_check(_, bufnr)
+  if vim.g.filetype_euphoria ~= nil then
```
```suggestion
  if vim.g.filetype_euphoria then
```
(and similarly everywhere)

clason commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -56,7 +87,15 @@ function M.ent(_, bufnr)
 end
 
-function M.ex_check(_, bufnr)
+function M.euphoria_check(_, bufnr)
```
```suggestion
function M.euphoria(_, bufnr)
```

smjonas committed:
Fixup: address review comments

smjonas committed:
Add functions for cfg, frm, header, idl, mc, mms and r filetypes

gpanders commented:
@smjonas Can you finish up porting the functions you've already started on and stop there? I don't want to port *all* of `dist#ft` in one go, as that will make reviewing the changes for correctness too difficult. We can do this in chunks in follow up PRs, which you are of course welcome to spearhead if you want to :)

clason commented:
Modest proposal: making the filetype detection _not_ depend on user settings like `ignorecase` is a feature? (As long as it is documented.)

smjonas committed:
Complete a few todos, finish tf function, always ignore case when =~ is used

smjonas commented:
@clason I agree that it should ignore the `ignorecase` option and instead always ignore case when `=~` is used (implicitly assuming `ignorecase = on` which is equivalent to `=~?`). I can't think of a case where that would detect the wrong filetype.

smjonas commented:
@gpanders Should I already update `filetype.lua` to use the new functions? Anyway, waiting for a quick review to address the TODOs, then it should be good to go for testing.

smjonas commented:
@gpanders Should I already update `filetype.lua` to use the new functions? Anyway, waiting for a quick review to address the TODOs, then it should be good to go for testing.

clason commented:
> @gpanders Should I already update `filetype.lua` to use the new functions? Anyway, waiting for a quick review to address the TODOs, then it should be good to go for testing.

Yes, if they're ported, they should be used (otherwise we can't test them ;))

smjonas committed:
Update filetype.lua

clason commented on runtime/lua/vim/filetype.lua:
```diff
@@ -780,19 +780,19 @@ local extension = {
-  EX = function() vim.fn["dist#ft#EuphoriaCheck"]() end,
-  EXU = function() vim.fn["dist#ft#EuphoriaCheck"]() end,
-  EXW = function() vim.fn["dist#ft#EuphoriaCheck"]() end,
+  EU = function(_, bufnr) return require("vim.filetype.detect").euphoria(_, bufnr) end,
```
```suggestion
  EU = function(path, bufnr) return require("vim.filetype.detect").euphoria(path, bufnr) end,
```

We want to keep this pattern consistent for now; can be refactored later when the porting is complete.

One thing I don't like is to _pass_  the underscore. I'd prefer if the `detect` function did not take unused arguments at all; the pattern would then be 

`EU = function(path, bufnr) return detect.euphoria(bufnr) end`

smjonas replied:
That seems like a good tradeoff, will do :+1:

smjonas committed:
Avoid passing _ to functions, remove unused arguments from functions in detect.lua

clason commented:
Looking at the BSD logs, I can already see some failures:

* `function RunTheTest[37]..Test_cfg_file line 19: Expected 'rapid' but got 'cfg'` (the `cfg` function is calling another function that is not yet ported)
* `function RunTheTest[37]..Test_ex_file line 9: Expected 'euphoria4' but got 'elixir'`
* `function RunTheTest[37]..Test_filetype_detection[2]..CheckItems line 13: with file name: file.r: Expected 'r' but got ''` (missing fallthrough?)
* `function RunTheTest[37]..Test_filetype_detection[2]..CheckItems line 13: with file name: file.tf: Expected 'tf' but got ''` (missing fallthrough?)

* And a bit of a worrying error:
```
Caught exception in Test_tf_file(): Vim(append):Error executing lua callback: ../../../runtime/lua/vim/filetype/detect.lua:18: Expected Lua numberstack traceback:	[C]: in function 'nvim_buf_get_lines'	../../../runtime/lua/vim/filetype/detect.lua:18: in function 'iter_lines'	../../../runtime/lua/vim/filetype/detect.lua:410: in function 'tf'	vim/filetype.lua:854: in function 'ft'	vim/filetype.lua:1562: in function 'dispatch'	vim/filetype.lua:1637: in function 'match'	/home/build/neovim/runtime/filetype.lua:15: in function </home/build/neovim/runtime/filetype.lua:14> @ BufReadPost Autocommands for "*"
```

clason commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,454 @@
+function M.nroff(bufnr)
+  for _, line in iter_lines(bufnr, 1, 5) do
+    if line:find("^%.") then
+      vim.bo[bufnr].filetype = "nroff"
```
This check is used elsewhere, so we need to mirror the return value (`1` if `nroff`, `0` else).

(Same for Perl below, when that gets ported.)

clason replied:
(Alternative -- forward looking -- would be to return `nroff` already and then just check `if detect.nroff() then ... end` at the calling site in `filetype.lua.)

smjonas committed:
Fix functions for cfg, ex, r and tf

Also rename iter_lines to get_lines and do not return an iterator because
that might lead to bugs when reusing the old iterator.
Rename get_lines to concat_lines.

clason commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -257,11 +256,13 @@ end
-  for _, line in iter_lines(bufnr, 1, 5) do
-    if line:find("^%.") then
+  for _, line in ipairs(get_lines(bufnr, 1, 5)) do
+    if line:sub(1, 1) == "." then
```
Why did you switch back from `find` to `sub`?

smjonas replied:
I thought it was more efficient in this case since we only need to check the first character. Do you prefer to use `find` for everything? I was only aware of the preference of `find` vs. `match`.

clason replied:
Yes, I think so -- unlike `sub`, `find` doesn't create a new string. 

In any case, I think we should be consistent for now; we can still switch en bloc later if benchmarks show that `sub` is in fact faster.

clason commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,456 @@
+end
+
+---@private
+local function concat_lines(bufnr, start_lnum, end_lnum)
```
I think we should combine these two functions by adding an `opts` table with a `concat` key (default whatever is simpler).

If default `false`, we can still special case if no `end_lnum` and `opts` is passed, in which case it just returns `nvim_buf_get_lines(bufnr, start_lnum-1, start_lnum, false)[1]` -- that should simplify the calling code as well.

smjonas committed:
Combine get_lines and concat_lines into a single function, fix tf

clason commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -6,15 +6,14 @@ local function count_lines(bufnr)
-  return vim.api.nvim_buf_get_lines(bufnr, start_lnum - 1, end_lnum, false)
-end
+local function get_lines(bufnr, start_lnum, end_lnum, opts)
+  if not end_lnum and not opts then
```
probably nobody would be silly enough to call `get_lines(0, 1, nil, opts)`, but just to be safe and cover all possibilities ;)
```suggestion
  if not end_lnum then
```

smjonas committed:
Update check in `detect.lua`

Co-authored-by: Christian Clason <christian.clason@uni-due.de>

smjonas committed:
Default opts table to {} to fix test failures, use concatenated lines in frm

smjonas committed:
Fix r()

gpanders commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,452 @@
+function M.cfg(path, bufnr)
+  if vim.g.filetype_cfg then
+    vim.bo[bufnr].filetype = vim.g.filetype_cfg
+  elseif M.is_rapid(path, bufnr) then
```
`IsRapid` is a "local" function in `dist#ft` so we can make it the same here (whenever you get to it).

gpanders commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,452 @@
+
+end
+
+function M.btm(bufnr)
```
✅

gpanders commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,452 @@
+-- This function checks for valid cl syntax in the first five lines.
+-- Look for either an opening comment, '#', or a block start, '{'.
+-- If not found, assume SGML.
+function M.ent(bufnr)
```
✅

gpanders commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,452 @@
+  vim.bo[bufnr].filetype = "dtd"
+end
+
+function M.euphoria(bufnr)
+  if vim.g.filetype_euphoria then
+    vim.bo[bufnr].filetype = vim.g.filetype_euphoria
+  else
+    vim.bo[bufnr].filetype = "euphoria3"
+  end
+end
```
✅

gpanders commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,452 @@
+    vim.bo[bufnr].filetype = vim.g.filetype_euphoria
+  else
+    for _, line in ipairs(get_lines(bufnr, 1, 100)) do
+    -- TODO: in the Vim regex, \> is used to match the end of the word, can this be omitted?
+      if line:find("^%-%-") or line:find("^ifdef") or line:find("^include")  then
```
```suggestion
      -- TODO: in the Vim regex, \> is used to match the end of the word, can this be omitted?
      if line:find("^%-%-") or line:find("^ifdef") or line:find("^include")  then
```

✅

gpanders commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,452 @@
+-- This function checks the first 15 lines for appearance of 'FoamFile'
+-- and then 'object' in a following line.
+-- In that case, it's probably an OpenFOAM file
+function M.foam(bufnr)
```
✅

gpanders commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,452 @@
+  end
+end
+
+function M.frm(bufnr)
```
✅

gpanders commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,452 @@
+
+end
+
+function M.header(bufnr)
```
✅

gpanders commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,452 @@
+  end
+end
+
+function M.idl(bufnr)
```
✅

gpanders commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,452 @@
+    vim.bo[bufnr].filetype = "abaqus"
+  else
+    for _, line in ipairs(get_lines(bufnr, 1, 500)) do
+      if line:find("^header surface data") then
```
`dist#ft#Check_inp()` uses `=~?`

```suggestion
      if line:lower():find("^header surface data") then
```

gpanders commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,452 @@
+
+-- Rely on the file to start with a comment.
+-- MS message text files use ';', Sendmail files use '#' or 'dnl'
+function M.mc(bufnr)
```
✅

gpanders commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,452 @@
+
+end
+
+function M.mms(bufnr)
```
✅

gpanders commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,452 @@
+
+-- This function checks if one of the first five lines start with a dot. In
+-- that case it is probably an nroff file: 'filetype' is set and 1 is returned.
+function M.nroff(bufnr)
```
✅

gpanders commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,452 @@
+  if vim.g.filetype_w then
+    vim.bo[bufnr].filetype = vim.g.filetype_w
+  else
+    if get_lines(bufnr, 1) == "&ANALYZE" or get_lines(bufnr, 3) == "&GLOBAL-DEFINE" then
```
```suggestion
    if get_lines(bufnr, 1):find("&ANALYZE") or get_lines(bufnr, 3):find("&GLOBAL%-DEFINE") then
```

gpanders commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,452 @@
+  local lines = get_lines(bufnr, 1, 50)
+  -- TODO: \< / \> which match the beginning / end of a word
+  -- Rebol is easy to recognize, check for that first
+  if table.concat(lines):find("[rR][eE][bB][oO][lL]") then
```
Might be easier/simpler to just lower case it first.

```suggestion
  if table.concat(lines):lower():find("rebel") then
```

gpanders commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,452 @@
+
+end
+
+function M.r(bufnr)
```
✅

gpanders commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,452 @@
+  end
+end
+
+function M.redif(bufnr)
```
✅

re: your TODO, I'd agree that using `lower()` is preferable for readability. If we can demonstrate a meaningful performance benefit hit from using it then we can revisit (but someone needs to do that benchmark).

gpanders commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,452 @@
+-- detection between scala and SuperCollider
+function M.sc(bufnr)
+  -- TODO: it still needs to be discussed if it's ok to use vim.regex in some cases
+  local regex = vim.regex([[[A-Za-z0-9]*\s:\s[A-Za-z0-9]\|var\s<\|classvar\s<\|\^this.*\||\w*|\|+\s\w*\s{\|\*ar\s]])
```
Do we need regex for this?

```lua
line:find("[A-Za-z0-9]*%s:%s[A-Za-z0-9]") or line:find("var%s<") or line:find("classvar%s<") or line:find("^this.*") or line:find("|%w*|") or line:find("%+%s%w*%s{") or line:find("%*ar%s")
```

It may be handy to define a helper function for these kinds of patterns. Something like

```lua
local function findany(s, pats)
    for _, v in ipairs(pats) do
        if s:find(v) then
            return true
        end
    end
    return false
end
```

and used like

```lua
findany(line, {"[A-Za-z0-9]*%s:%s[A-Za-z0-9]", "var%s<", "classvar%s<", "^this.*", "|%w*|", "%+%s%w*%s{", "%*ar%s"})
```

clason replied:
Sounds like a good idea; that will crop up a few more times (also in `scripts.vim`, which I'd like to add -- later -- as `detect.scripts()`).

gpanders commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,452 @@
+-- detection between scdoc and SuperCollider
+function M.scd(bufnr)
+  -- TODO: it still needs to be discussed if it's ok to use vim.regex in some cases
+  local regex = vim.regex([[\%^\S\+(\d[0-9A-Za-z]*)\%(\s\+\"[^"]*\"\%(\s\+\"[^"]*\"\)\=\)\=$]])
```
This one's a bit ugly with Lua patterns as we have to repeat ourselves a few times, but it's technically possible.

```lua
local first = "^%S+%(%d[0-9A-Za-z]*%)"
local opt = "%s+\"[^\"]*\""
line:find(first .. "$") or line:find(first .. opt .. "$") or line:find(first .. opt .. opt .. "$")
```

gpanders commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,452 @@
+
+end
+
+function M.sql(bufnr)
```
✅

gpanders commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,452 @@
+end
+
+-- Determine if a *.tf file is TF mud client or terraform
+function M.tf(bufnr)
```
✅

gpanders commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,452 @@
+  vim.bo[bufnr].filetype = "tf"
+end
+
+function M.xml(bufnr)
```
✅

gpanders commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,452 @@
+  vim.bo[bufnr].filetype = "xml"
+end
+
+function M.y(bufnr)
```
✅

smjonas committed:
Add findany helper function, replace vim regexes with lua patterns

clason commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,397 @@
+  end
+end
+
+local function is_rapid(path, bufnr) end
```
Shouldn't forget to actually port this function since it's used above ;)

clason commented:
I'm leaving a comment here so I won't forget (need not be done in this PR): should we embed `vim.filetype.detect` as bytecode during build?

smjonas committed:
Implement is_rapid which is used in cfg

smjonas commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -43,10 +43,27 @@ function M.btm(bufnr)
+  -- Called from mod, prg or sys functions
+  for _, line in ipairs(getlines(bufnr, 1, -1)) do
+    if not line:find("^%s*$") then
+      return findany(line:lower(), { "^%s*%%%%%%", first .. "(", first .. "$" })
```
@gpanders I hope it's fine that I implemented this test function in this PR ;) Do you think I understood `\%{3}` correctly in the Vim regex: (`\v^\s*%(\%{3}|module\s+\k+\s*%(\(|$))`)? I think these are 3 percentage signs but not 100% sure.

gpanders replied:
>Do you think I understood \%{3} correctly in the Vim regex: (\v^\s*%(\%{3}|module\s+\k+\s*%(\(|$)))? I think these are 3 percentage signs but not 100% sure.

Yes that is correct.

smjonas commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -43,10 +43,27 @@ function M.btm(bufnr)
+    local line = getlines(bufnr, 1):lower()
+    return findany(line, { "eio:cfg", "mmc:cfg", "moc:cfg", "proc:cfg", "sio:cfg", "sys:cfg" })
+  end
+  local first = "^%s*module%s+%S+%s*"
```
Is ignoring the keyword argument ok in favor of `%S`?

gpanders replied:
This is tricky. `\k` might be too difficult to translate into pure Lua, so this may be a case where `vim.regex` is warranted.

clason commented:
I think now is a good time to squash your commits (and make commitlint happy ;))

clason commented:
(Might also want to rebase on master for good measure.)

clason commented:
Luacheck needs directives to shut up about some things -- you can steal them from `filetype.lua`.

smjonas commented:
I tried squashing using `git rebase -i 2caf5bbb` but GitHub rejects the force push?
```
! [remote rejected]     filetype-detect -> filetype-detect (bad ref)                                                                                                                                                                   
error: failed to push some refs to 'git@github.com:smjonas/neovim.git'
```

 I also tried with `git reset --soft 2caf5bbb` before. I am bit confused since it worked the first time I squashed this PR.

clason commented:
Might be partial Github outage.

clason commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,417 @@
+    if
+      findany(
+        line,
+        { "[A-Za-z0-9]*%s:%s[A-Za-z0-9]", "var%s<", "classvar%s<", "^this.*", "|%w*|", "%+%s%w*%s{", "%*ar%s" }
```
Test indicates that something here is not equivalent to the vim version; it's failing to recognize
```
// Query
//Method
^this {
```
as `supercollider`

I think this is meant to be a _literal_ `^`?

smjonas committed:
Change ^ to literal ^ in sc

clason commented:
Hmm, I couldn't really say -- I never had these issues (unless Github itself was acting up), and squash and force-push all the time...

But it's not critical; we can squash through Github when merging!

smjonas commented:
Perfect, feel free to do that!

clason commented:
Tests look good (CI failures are expected flake), but the Lua linter still complains. You need to
1. remove the `countlines` function and
2. copy the `-- luacheck: push ...` line from `filetype.lua` (and possibly add more)

smjonas committed:
Fix luacheck errors

smjonas commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -409,4 +407,7 @@ function M.y(bufnr)
   vim.bo[bufnr].filetype = "yacc"
 end
 
+-- luacheck: pop
```
Not sure if push / pop is necessary or if
```
-- luacheck: push no unused args
-- luacheck: push ignore 122
```
would be enough.

clason replied:
I think `pop` is correct -- we do that in `filetype.lua` as well.

smjonas replied:
Yes its correct but only required when the errors should not be ignored for the whole file (only up to the pop statement).

clason replied:
True, but still good practice :)

clason commented:
And lint is happy, too :)

LGTM! @gpanders would you give it a final review?

gpanders commented:
Let's go for it. Thanks @smjonas for all your hard work on this!

clason commented:
Thank you @smjonas for all the work -- and your patience!

If you want to continue porting further autoload functions, or `scripts.vim`, that would be more than welcome!

smjonas commented:
Thank you both for helping me with this PR and your useful feedback, it was great working with you! Sure, I'll definitely consider porting more stuff in the future! 10 / 10 - would port again :smile:

ranjithshegde commented on runtime/lua/vim/filetype/detect.lua:
```diff
@@ -0,0 +1,413 @@
+
+-- This function checks the first line of file extension "scd" to resolve
+-- detection between scdoc and SuperCollider
+function M.scd(bufnr)
```
For some reason it always defaults to scdoc even if the pattern is not a match or a blank file. never sets "supercollider"
Will reproduce with minimal init.lua and file an issue

clason replied:
It passes the test though,which checks for supercollider. Make sure impatient is not caching something.

gpanders replied:
Yes it looks like there is a bug in the `scd` detection:

https://github.com/neovim/neovim/blob/9005ffbe7757eca8ad809c81db76aec930db8e68/runtime/lua/vim/filetype/detect.lua#L338

The first entry `"$"` should not be there.

gpanders replied:
@clason Alas, `test_filetype.vim` only checks for `supercollider` on `.sc` files. For `.scd` files it only checks for `scdoc`, which explains why the test didn't catch it.

clason replied:
oops, sorry, I misread that! (was on mobile...)

@ranjithshegde thanks for catching that!

ranjithshegde replied:
@gpanders your merge solves this! thanks
